feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/handlers/feeding/DefaultFeedingHandler.kt

```diff
@@ -1,43 +1,103 @@
-    private val repository: FeedingPointRepository,
-    private val saveUserAsFeederUseCase: SaveUserAsFeederUseCase,
-    stateDelegate: StateDelegate<HomeState>
-) : FeedingHandler, StateDelegate<HomeState> by stateDelegate {
-
-    override suspend fun fetchFeedingPoints() {
-        repository.getAllFeedingPoints().collect {
-            updateState {
-                copy(
-                    feedingPoints = it.map { feedingPoint ->
-                        FeedingPointModel(feedingPoint)
-                    }.toImmutableList()
-                )
-            }
+@Suppress("LongParameterList")
+internal class DefaultFeedingHandler @Inject constructor(
+    stateDelegate: StateDelegate<HomeState>,
+    routeHandler: RouteHandler,
+    errorHandler: ErrorHandler,
+    feedingPointHandler: FeedingPointHandler,
+    private val startFeedingUseCase: StartFeedingUseCase,
+    private val cancelFeedingUseCase: CancelFeedingUseCase,
+    private val finishFeedingUseCase: FinishFeedingUseCase,
+    private val dispatchers: Dispatchers
+) : FeedingHandler,
+    StateDelegate<HomeState> by stateDelegate,
+    FeedingPointHandler by feedingPointHandler,
+    RouteHandler by routeHandler,
+    ErrorHandler by errorHandler {
+
+    override fun CoroutineScope.handleFeedingEvent(event: FeedingEvent) {
+        when (event) {
+            Start -> launch { startFeeding() }
+            Cancel -> launch { cancelFeeding() }
+            Finish -> launch { finishFeeding() }
         }
     }
 
-    override fun hideFeedingPointsButOne(feedingPoint: FeedingPointModel) {
-        updateState {
-            copy(feedingPoints = persistentListOf(feedingPoint))
-        }
+    private suspend fun startFeeding() {
+        performFeedingAction(
+            action = startFeedingUseCase::invoke,
+            onSuccess = { currentFeedingPoint ->
+                hideOtherFeedingPoints(currentFeedingPoint)
+                startRouteAndTimer()
+            }
+        )
     }
 
-    override fun saveFeeder(): Flow<Boolean> {
-        return saveUserAsFeederUseCase(
-            state.currentFeedingPoint?.id ?: return flow { emit(false) }
+    private suspend fun cancelFeeding() {
+        performFeedingAction(
+            action = cancelFeedingUseCase::invoke,
+            onSuccess = {
+                stopRouteAndTimer()
+                fetchFeedingPoints()
+            }
         )
     }
+
+    private suspend fun finishFeeding() {
+        performFeedingAction(
+            action = { feedingPointId ->
+                finishFeedingUseCase(feedingPointId, listOf(""))
+            },
+            onSuccess = {
+                stopRouteAndTimer()
+                fetchFeedingPoints()
+            }
+        )
+    }
+
+    private suspend fun performFeedingAction(
+        action: suspend (String) -> ActionResult,
+        onSuccess: suspend (FeedingPointModel) -> Unit
+    ) {
+        state.currentFeedingPoint?.let { currentFeedingPoint ->
```

Comment:
I suppose most of this can be moved to util for requests that return ActionResult since it is not explicitly tied to feeding only and might be reused later.

Reply:
Good point. Moved to `ActionHandler` in `common`